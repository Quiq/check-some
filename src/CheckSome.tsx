import React from 'react';
import * as PropTypes from 'prop-types';
import isEqual from 'lodash/isEqual';
import CheckSomeField from './Field';
import {CHECK_SOME_CONTEXT} from './globals';
import {ValidationErrors} from './globals';

type ValidationRule = (value: any) => ValidationErrors | null;

type ValidationGroupRules = {[name: string]: Array<ValidationRule>};
type ValidationGroupErrors = {[name: string]: ValidationErrors} | null;

export type CheckSomeChildProps = {
  valid: boolean,
  changed: boolean,
  errors: ValidationGroupErrors,
};

export type CheckSomeProps = {
  rules: ValidationGroupRules,
  values: {[key: string]: any}, // TODO: Get a better type here
  initialValues?: {[key: string]: any},
  children: (props: CheckSomeChildProps) => React.ReactNode,
};

/* eslint-disable react/no-multi-comp */
export default class CheckSome extends React.Component<CheckSomeProps> {
  static Field = CheckSomeField;
  static childContextTypes = {
    [CHECK_SOME_CONTEXT]: PropTypes.object.isRequired,
  };

  initialValues: Object | null | undefined;

  getChildContext() {
    return {
      [CHECK_SOME_CONTEXT]: {
        values: this.props.values,
        errors: this.getErrors(),
      },
    };
  }

  getInitialValues = () => {
    if (this.props.initialValues) {
      return this.props.initialValues;
    }

    if (!this.initialValues) {
      this.initialValues = this.props.values;
    }

    return this.initialValues;
  };

  getErrors = (): ValidationGroupErrors =>
    Object.keys(this.props.rules).reduce((errors: ValidationGroupErrors, key) => {
      const rules = this.props.rules[key];
      const value = this.props.values[key];

      const newErrors = rules.reduce(
        (e: ValidationErrors | null, rule: ValidationRule) => e || rule(value),
        null,
      );

      if (!newErrors) {
        return errors;
      }
      if (!errors) {
        // Initialize error object if this is the first error
        errors = {};
      }

      errors[key] = newErrors;

      return errors;
    }, null);

  render() {
    const {values} = this.props;

    const errors = this.getErrors();
    const valid = !errors;

    const changed = !isEqual(values, this.getInitialValues());

    return this.props.children({valid, errors, changed});
  }
}
