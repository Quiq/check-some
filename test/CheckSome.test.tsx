import React from 'react';
import CheckSome from '../src';
import {render, fireEvent, RenderResult} from '@testing-library/react';

const required = value => (value || value === 0 ? null : {required: {}});

const greaterThanZero = (value: number) => (value > 0 ? null : {greaterThanZero: {value}});

const TestField = ({value, errors, valid, touched}) => (
  <div className="field">
    <input defaultValue={value} />
    <div>Field {valid ? 'Valid' : 'Invalid'}</div>
    <div>{touched ? 'Touched' : 'Pristine'}</div>
    <div>Field Errors: {JSON.stringify(errors)}</div>
  </div>
);

const TestForm = ({valid, changed, errors}) => (
  <form>
    <div>Form {valid ? 'Valid' : 'Invalid'}</div>
    <div>{changed ? 'Changed' : 'Unchanged'}</div>
    <div>Form Errors: {JSON.stringify(errors)}</div>

    <CheckSome.Field name="requiredString">
      {fieldProps => <TestField {...fieldProps} />}
    </CheckSome.Field>

    <CheckSome.Field name="testNumber">
      {fieldProps => <TestField {...fieldProps} />}
    </CheckSome.Field>

    <CheckSome.Field name="optionalString">
      {fieldProps => <TestField {...fieldProps} />}
    </CheckSome.Field>
  </form>
);

describe('CheckSome', () => {
  let r: RenderResult;
  let testProps;

  beforeEach(() => {
    const values = {
      requiredString: '',
      testNumber: undefined,
      optionalString: '',
    };

    const rules = {
      requiredString: [required],
      testNumber: [required, greaterThanZero],
    };

    testProps = {values, rules};

    r = render(<CheckSome {...testProps}>{formProps => <TestForm {...formProps} />}</CheckSome>);
  });

  it('renders the initial form', () => {
    expect(r.container).toMatchSnapshot();
    expect(r.getAllByText('Pristine').length).toBe(3);
  });

  describe('blurring the fields', () => {
    beforeEach(() => {
      r.container.querySelectorAll('input').forEach(i => fireEvent.blur(i));
    });

    it('sets the touched prop for fields', () => {
      expect(r.container).toMatchSnapshot();
      expect(r.getAllByText('Touched').length).toBe(3);
    });
  });

  describe('setting values to something still invalid', () => {
    beforeEach(() => {
      testProps.values = {
        requiredString: 'spongebob',
        testNumber: -2,
        optionalString: 'patrick',
      };
      r.rerender(<CheckSome {...testProps}>{formProps => <TestForm {...formProps} />}</CheckSome>);
    });

    it('updates changed, values, and errors', () => {
      expect(r.container).toMatchSnapshot();
    });
  });

  describe('setting values to something valid', () => {
    beforeEach(() => {
      testProps.values = {
        requiredString: 'spongebob',
        testNumber: 7,
        optionalString: 'patrick',
      };
      r.rerender(<CheckSome {...testProps}>{formProps => <TestForm {...formProps} />}</CheckSome>);
    });

    it('updates changed, values, and errors', () => {
      expect(r.container).toMatchSnapshot();
    });
  });
});
