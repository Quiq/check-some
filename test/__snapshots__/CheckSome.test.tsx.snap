// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`CheckSome blurring the fields sets the touched prop for fields 1`] = `
<div>
  <form>
    <div>
      Form 
      Invalid
    </div>
    <div>
      Unchanged
    </div>
    <div>
      Form Errors: 
      {"requiredString":{"required":{}},"testNumber":{"required":{}}}
    </div>
    <div
      class="CheckSomeField"
    >
      <div
        class="field"
      >
        <input
          value=""
        />
        <div>
          Field 
          Invalid
        </div>
        <div>
          Touched
        </div>
        <div>
          Field Errors: 
          {"required":{}}
        </div>
      </div>
    </div>
    <div
      class="CheckSomeField"
    >
      <div
        class="field"
      >
        <input
          value=""
        />
        <div>
          Field 
          Invalid
        </div>
        <div>
          Touched
        </div>
        <div>
          Field Errors: 
          {"required":{}}
        </div>
      </div>
    </div>
    <div
      class="CheckSomeField"
    >
      <div
        class="field"
      >
        <input
          value=""
        />
        <div>
          Field 
          Valid
        </div>
        <div>
          Touched
        </div>
        <div>
          Field Errors: 
        </div>
      </div>
    </div>
  </form>
</div>
`;

exports[`CheckSome renders the initial form 1`] = `
<div>
  <form>
    <div>
      Form 
      Invalid
    </div>
    <div>
      Unchanged
    </div>
    <div>
      Form Errors: 
      {"requiredString":{"required":{}},"testNumber":{"required":{}}}
    </div>
    <div
      class="CheckSomeField"
    >
      <div
        class="field"
      >
        <input
          value=""
        />
        <div>
          Field 
          Invalid
        </div>
        <div>
          Pristine
        </div>
        <div>
          Field Errors: 
          {"required":{}}
        </div>
      </div>
    </div>
    <div
      class="CheckSomeField"
    >
      <div
        class="field"
      >
        <input
          value=""
        />
        <div>
          Field 
          Invalid
        </div>
        <div>
          Pristine
        </div>
        <div>
          Field Errors: 
          {"required":{}}
        </div>
      </div>
    </div>
    <div
      class="CheckSomeField"
    >
      <div
        class="field"
      >
        <input
          value=""
        />
        <div>
          Field 
          Valid
        </div>
        <div>
          Pristine
        </div>
        <div>
          Field Errors: 
        </div>
      </div>
    </div>
  </form>
</div>
`;

exports[`CheckSome setting values to something still invalid updates changed, values, and errors 1`] = `
<div>
  <form>
    <div>
      Form 
      Invalid
    </div>
    <div>
      Changed
    </div>
    <div>
      Form Errors: 
      {"testNumber":{"greaterThanZero":{"value":-2}}}
    </div>
    <div
      class="CheckSomeField"
    >
      <div
        class="field"
      >
        <input
          value="spongebob"
        />
        <div>
          Field 
          Valid
        </div>
        <div>
          Pristine
        </div>
        <div>
          Field Errors: 
        </div>
      </div>
    </div>
    <div
      class="CheckSomeField"
    >
      <div
        class="field"
      >
        <input
          value="-2"
        />
        <div>
          Field 
          Invalid
        </div>
        <div>
          Pristine
        </div>
        <div>
          Field Errors: 
          {"greaterThanZero":{"value":-2}}
        </div>
      </div>
    </div>
    <div
      class="CheckSomeField"
    >
      <div
        class="field"
      >
        <input
          value="patrick"
        />
        <div>
          Field 
          Valid
        </div>
        <div>
          Pristine
        </div>
        <div>
          Field Errors: 
        </div>
      </div>
    </div>
  </form>
</div>
`;

exports[`CheckSome setting values to something valid updates changed, values, and errors 1`] = `
<div>
  <form>
    <div>
      Form 
      Valid
    </div>
    <div>
      Changed
    </div>
    <div>
      Form Errors: 
      null
    </div>
    <div
      class="CheckSomeField"
    >
      <div
        class="field"
      >
        <input
          value="spongebob"
        />
        <div>
          Field 
          Valid
        </div>
        <div>
          Pristine
        </div>
        <div>
          Field Errors: 
          null
        </div>
      </div>
    </div>
    <div
      class="CheckSomeField"
    >
      <div
        class="field"
      >
        <input
          value="7"
        />
        <div>
          Field 
          Valid
        </div>
        <div>
          Pristine
        </div>
        <div>
          Field Errors: 
          null
        </div>
      </div>
    </div>
    <div
      class="CheckSomeField"
    >
      <div
        class="field"
      >
        <input
          value="patrick"
        />
        <div>
          Field 
          Valid
        </div>
        <div>
          Pristine
        </div>
        <div>
          Field Errors: 
          null
        </div>
      </div>
    </div>
  </form>
</div>
`;
